packages/adminrouter/extra/src/test-harness/tests/test_service.py

```diff
@@ -0,0 +1,575 @@
+        # Make MesosDNS mock respond with garbled data
+        mocker.send_command(endpoint_id='http://127.0.0.1:8123',
+                            func_name='set_encoded_response',
+                            aux_data=b'blah blah duh duh')
+
+        # Verify the response:
+        url = master_ar_process_fastcache.make_url_from_path(
+            '/service/scheduler-alwaysthere/foo/bar/')
+        resp = requests.get(url,
+                            allow_redirects=False,
+                            headers=valid_user_header)
+
+        assert resp.status_code == 503
+
+    def test_if_broken_response_status_from_mesos_dns_is_handled(
+            self, master_ar_process_fastcache, mocker, valid_user_header):
+        # Remove the data from Mesos and Marathon mocks w.r.t. resolved service
+        mocker.send_command(endpoint_id='http://127.0.0.1:8080',
+                            func_name='set_apps_response',
+                            aux_data={"apps": []})
+        mocker.send_command(endpoint_id='http://127.0.0.2:5050',
+                            func_name='set_frameworks_response',
+                            aux_data=[SCHEDULER_FWRK_ALWAYSTHERE_NOWEBUI])
+
+        # Make MesosDNS mock respond with invalid data
+        mocker.send_command(endpoint_id='http://127.0.0.1:8123',
+                            func_name='always_bork',
+                            aux_data=True)
+
+        # Verify the response:
+        url = master_ar_process_fastcache.make_url_from_path(
+            '/service/scheduler-alwaysthere/foo/bar/')
+        resp = requests.get(url,
+                            allow_redirects=False,
+                            headers=valid_user_header)
+
+        assert resp.status_code == 503
+
+    def test_if_timed_out_response_from_mesos_dns_is_handled(
+            self, master_ar_process_fastcache, mocker, valid_user_header):
+        # Remove the data from Mesos and Marathon mocks w.r.t. resolved service
+        mocker.send_command(endpoint_id='http://127.0.0.1:8080',
+                            func_name='set_apps_response',
+                            aux_data={"apps": []})
+        mocker.send_command(endpoint_id='http://127.0.0.2:5050',
+                            func_name='set_frameworks_response',
+                            aux_data=[SCHEDULER_FWRK_ALWAYSTHERE_NOWEBUI])
+
+        # Make MesosDNS mock stall response by 10s
+        mocker.send_command(endpoint_id='http://127.0.0.1:8123',
+                            func_name='always_stall',
+                            aux_data=10)
+
+        # Verify the response:
+        url = master_ar_process_fastcache.make_url_from_path(
+            '/service/scheduler-alwaysthere/foo/bar/')
+        t_start = time.time()
+        resp = requests.get(url,
+                            allow_redirects=False,
+                            headers=valid_user_header)
+        t_spent = time.time() - t_start
+
+        assert t_spent < 10
```

Comment:
```
# If the timeout was properly enforced by Admin Router, the total time
# spent waiting for response will be less than 10s. If there is no
# timeout - it will be at least 10s.
```